cli/client.go

```diff
@@ -0,0 +1,305 @@
+	}
+
+	var alts []*model.Alert
+	err = json.Unmarshal(body, &alts)
+
+	return alts, err
+}
+
+func (h *httpAlertAPI) Push(ctx context.Context, alerts ...model.Alert) error {
+	u := h.client.URL(epAlerts, nil)
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(&alerts); err != nil {
+		return err
+	}
+
+	req, _ := http.NewRequest(http.MethodPost, u.String(), &buf)
+
+	_, _, err := h.client.Do(ctx, req)
+	return err
+}
+
+// SilenceAPI provides bindings for the Alertmanager's silence API.
+type SilenceAPI interface {
+	// Get returns the silence associated with the given ID.
+	Get(ctx context.Context, id string) (*types.Silence, error)
+	// Set updates or creates the given silence and returns its ID.
+	Set(ctx context.Context, sil types.Silence) (string, error)
+	// Delete deletes the silence with the given ID.
+	Delete(ctx context.Context, id string) error
+	// List returns all the silences.
+	List(ctx context.Context) ([]*types.Silence, error)
+}
+
+// NewSilenceAPI returns a new SilenceAPI for the client.
+func NewSilenceAPI(c api.Client) SilenceAPI {
+	return &httpSilenceAPI{client: apiClient{c}}
+}
+
+type httpSilenceAPI struct {
+	client api.Client
+}
+
+func (h *httpSilenceAPI) Get(ctx context.Context, id string) (*types.Silence, error) {
+	u := h.client.URL(epSilence, map[string]string{
+		"id": id,
+	})
+
+	req, _ := http.NewRequest(http.MethodGet, u.String(), nil)
+
+	_, body, err := h.client.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var sil types.Silence
+	err = json.Unmarshal(body, &sil)
+
+	return &sil, err
+}
+
+func (h *httpSilenceAPI) Delete(ctx context.Context, id string) error {
+	u := h.client.URL(epSilence, map[string]string{
+		"id": id,
+	})
+
+	req, _ := http.NewRequest(http.MethodDelete, u.String(), nil)
+
+	_, _, err := h.client.Do(ctx, req)
+	return err
+}
+
+func (h *httpSilenceAPI) Set(ctx context.Context, sil types.Silence) (string, error) {
+	var (
+		u      *url.URL
+		method string
+	)
+
+	var buf bytes.Buffer
+	s := types.Silence{
+		Matchers:  make([]*types.Matcher, len(sil.Matchers)),
+		StartsAt:  sil.StartsAt,
+		EndsAt:    sil.EndsAt,
+		CreatedBy: sil.CreatedBy,
+		Comment:   sil.Comment,
+	}
+	for _, v := range sil.Matchers {
+		s.Matchers = append(s.Matchers, &types.Matcher{Name: string(v.Name), Value: v.Value, IsRegex: v.IsRegex})
+	}
+	if err := json.NewEncoder(&buf).Encode(&s); err != nil {
+		return "", err
+	}
+
+	u = h.client.URL(epSilences, nil)
+	method = http.MethodPost
+
+	req, _ := http.NewRequest(method, u.String(), &buf)
+
+	_, body, err := h.client.Do(ctx, req)
+	if err != nil {
+		return "", err
+	}
+
+	var res struct {
+		SilenceID string `json:"silenceId"`
+	}
+	err = json.Unmarshal(body, &res)
+
+	return res.SilenceID, err
+}
+
+func (h *httpSilenceAPI) List(ctx context.Context) ([]*types.Silence, error) {
+	u := h.client.URL(epSilences, nil)
+
+	req, _ := http.NewRequest(http.MethodGet, u.String(), nil)
+
+	_, body, err := h.client.Do(ctx, req)
```

Comment:
it would be good to check the response code. A `404` or `500` is still a successful http response, but something we want to inform the api consumers of.

Reply:
This is a call to the local `apiClient.Do()` which already handles the HTTP response code and un-marshals the server's response.